### PR TITLE
Simplify logic

### PR DESCRIPTION
**Pull request by Ragil**

### CODE REVIEW

File: app/api/sms/plant_action.py [hunk 1]
```diff
@@ -16,26 +16,24 @@ def __init__(self, command):
...
-            update.quantity = plant.quantity + cmd.amount
-            update.put()
-        else:
-            new = Farm(id=Farm.id(),
-                       district_id=user.district_id,
-                       action=self.CMD,
-                       crop_name=cmd.plant,
-                       quantity=cmd.amount)
-            new.put()
+                                   Farm.action == 'plant')).get()
```
Comment (Ragil) [inline app/api/sms/plant_action.py]: A simpler api to fetch 1 item.

https://cloud.google.com/appengine/docs/python/ndb/queryclass#Query_get

File: app/api/sms/plant_action.py [hunk 2]
```diff
@@ -16,26 +16,24 @@ def __init__(self, command):
...
+        if not plant:
+            plant = Farm(id=Farm.id(),
```
Comment (Ragil) [inline app/api/sms/plant_action.py]: Reducing the if else tree makes the logic easier to understand.

File: app/api/sms/query_action.py [hunk 3]
```diff
@@ -28,14 +23,12 @@ def execute(self):
...
         filter = self.command.filter
-        if filter in self.FILTER:
-            filter = self.FILTER[self.command.filter]
 
         query = Farm.query(ndb.AND(Farm.action == filter,
                                    Farm.district_id == district_id))
         crops = query.order(-Farm.ts_updated).fetch(self.LIMIT)
 
-        if len(crops) == 0:
+        if not crops:
```
Comment (Ragil) [inline app/api/sms/query_action.py]: In python, an empty list is considered False. https://docs.python.org/2/library/stdtypes.html#truth-value-testing

File: app/api/sms/query_action.py [hunk 4]
```diff
@@ -28,14 +23,12 @@ def execute(self):
         district_id = district.key.id()
 
         filter = self.command.filter
-        if filter in self.FILTER:
-            filter = self.FILTER[self.command.filter]
```
Comment (Ragil) [inline app/api/sms/query_action.py]: Parsing the sms body into the exact format that an action takes is the responsibility of the command. The action simply performs the action it is told to do.

File: tests/api/sms/test_broadcast_old.py [hunk 5]
```diff
@@ -12,6 +12,7 @@
 from google.appengine.ext import testbed, ndb
 
 
+@unittest.skip('need to be refactored')
```
Comment (Ragil) [inline tests/api/sms/test_broadcast_old.py]: These tests are currently skipped. They should be broken down into smaller tests that verify each component independently of each other.

File: tests/api/sms/test_harvest_command.py [hunk 6]
```diff
@@ -77,39 +74,3 @@ def test_harvest_command_with_invalid_command(self):
             self.sms.body = body
             cmd = HarvestCommand(self.sms)
             self.assertFalse(cmd.valid())
-
-    @patch('app.api.sms.main.dispatcher')
-    def test_harvest_action(self, mock):
```
Comment (Ragil) [inline tests/api/sms/test_harvest_command.py]: This test does far too many things. It tries to replicate the dispatcher. It tests that the command can handle multiple different formats. It tests we can parse the twilio sms request.

Unfortunately, this does not actually test the harvest action very well. It does not test all the edge cases associated to the harvest action.

This has been re factored appropriately into https://github.com/Ragil/Communicate-Indonesia/pull/48/files#diff-84d7355c64a73e697de156a99a362a20R10

File: app/api/sms/harvest_action.py [hunk 7]
```diff
@@ -58,17 +66,17 @@ class HarvestCommand(ThreeArgCommand):
...
     def __init__(self, sms):
         super(HarvestCommand, self).__init__(sms)
-        self.harvest = None
+        self.plant = None
         self.amount = None
 
         if self.args[0] and self.args[0].isdigit():
             self.amount = int(self.args[0])
-            self.harvest = self.args[1]
+            self.plant = self.args[1]
```
Comment (Ragil) [inline app/api/sms/harvest_action.py]: Rename because harvest acts on a plant

Comment (Ragil) [inline app/api/sms/harvest_action.py]: Technically, its a crop.

File: tests/api/test_sms_api.py [hunk 8]
```diff
@@ -114,20 +62,32 @@ def test_multi_harvest_empty_sell_plant(self):
...
+        res = self.send('look harvest sumatra')
         self.assertTrue('Total panen di Sumatra:'
-                        '\nWortel 3' in res.data)
+                        '\nKentang 3' in res.data)
+
+        res = self.send('look sumatra')
+        # TODO - the sell action should subtract from harvested total
+        #      - Fix the bug in sell action for this assertion to work
+        # self.assertTrue('Total panen di Sumatra:'
+        #                '\nKentang 2' in res.data)
```
Comment (Ragil) [inline tests/api/test_sms_api.py]: This should be uncommented after sell action is fixed.

File: tests/api/sms/test_harvest_command.py [hunk 9]
```diff
@@ -74,42 +55,7 @@ def test_harvest_command_with_invalid_command(self):
         ]
 
         for body in invalid_messages:
-            self.sms.body = body
-            cmd = HarvestCommand(self.sms)
-            self.assertFalse(cmd.valid())
-
-    @patch('app.api.sms.main.dispatcher')
-    def test_harvest_action(self, mock):
-        mock.dispatch.return_value = None
```
Comment (Ragil) [inline tests/api/sms/test_harvest_command.py]: This test does far too many things. It tries to replicate the dispatcher. It tests that the command can handle multiple different formats. It tests we can parse the twilio sms request.

Unfortunately, this does not actually test the harvest action very well. It does not test all the edge cases associated to the harvest action.

This has been re factored appropriately into https://github.com/Ragil/Communicate-Indonesia/pull/48/files#diff-84d7355c64a73e697de156a99a362a20R10

File: tests/api/test_sms_api.py [hunk 10]
```diff
@@ -35,55 +29,13 @@ def tearDown(self):
...
-                        '\nKentang 5' in res.data)
-
-    def test_single_harvest(self):
-        res = self.send('harvest potato 2')
-        self.assertTrue(_('Harvest command succeeded') in res.data)
-
-        res = self.send('look sumatra')
-        self.assertTrue('Total panen di Sumatra:'
-                        '\nKentang 2' in res.data)
-
```
Comment (Ragil) [inline tests/api/test_sms_api.py]: Each of these have been tested in the appropriate command/action tests. The integration of these commands are tested in the two test cases below that exercise the general case.
